jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java

```diff
@@ -0,0 +1,124 @@
+import javax.enterprise.context.spi.Context;
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.AfterTypeDiscovery;
+import javax.enterprise.inject.spi.AnnotatedType;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.Extension;
+
+import org.hibernate.search.jsr352.context.jpa.EntityManagerFactoryRegistry;
+import org.hibernate.search.jsr352.inject.scope.HibernateSearchJobScoped;
+import org.hibernate.search.jsr352.inject.scope.HibernateSearchPartitionScoped;
+import org.hibernate.search.jsr352.massindexing.impl.JobContextSetupListener;
+import org.hibernate.search.jsr352.massindexing.impl.steps.lucene.EntityReader;
+import org.jberet.cdi.JobScoped;
+import org.jberet.cdi.PartitionScoped;
+
+
+/**
+ * A CDI extension to explicitly add hibernate-search-jsr352-core types to the CDI context.
+ * <p>
+ * These types must be registered into the CDI context because they expect one of their
+ * attributes to be injected, thus they need to be instantiated by CDI.
+ * <p>
+ * Ideally we should use component scanning to that effect, but:
+ * <ul>
+ * <li>In CDI, component scanning is configured per-module in META-INF/beans.xml,
+ * with no way to add a package from another module.
+ * <li>In Wildfly, component scanning in dependencies is always done with the "all" discovery type,
+ * meaning that even components without CDI annotations will be discovered.
+ * Besides exposing private types unnecessarily,
+ * this is annoying because it puts default {@link EntityManagerFactoryRegistry} implementations into the CDI context,
+ * leading to conflicts when injecting such types
+ * (since this module, hibernate-search-jsr352-jberet, also provides an implementation).
+ * <br>See <a href="https://issues.jboss.org/browse/WFLY-8656">WFLY-8656</a>.
+ * <li>Even if we could solve the above, JBeret requires specific scope annotations to be set on the
+ * beans so that they are correctly scoped. Since the core isn't dependent on JBeret, we must specify these
+ * scopes manually when registering the beans.
+ * </ul>
+ * Thus we use explicit type and scope registration as a workaround.
+ *
+ * @author Yoann Rodiere
+ */
+public class HibernateSearchJsr352Extension implements Extension {
+
+	public void afterTypeDiscovery(@Observes AfterTypeDiscovery event, BeanManager beanManager) {
+		registerType( event, beanManager, JobContextSetupListener.class );
+		registerType( event, beanManager, EntityReader.class );
+	}
+
+	public void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager beanManager) {
+		addScopeAlias( event, beanManager, HibernateSearchJobScoped.class, JobScoped.class );
+		addScopeAlias( event, beanManager, HibernateSearchPartitionScoped.class, PartitionScoped.class );
```

Comment:
> Here, we add manually a scope alias after the bean discovery. Or, in other words, after starting a batch job execution

Actually this code is executed when CDI starts, which is only once and happens before a batch job is even started. But the alias will be effective for any job. We don't manage the scope lifecycle here, we just put some scope aliases; the actual scope lifecycle implementation is part of JBeret.

Apart from that, what you're correct.